lib/commands/context.js

```diff
@@ -0,0 +1,462 @@
+  if (alreadyInContext(name, this.curContext)) {
+    // already in the named context, no need to do anything
+  } else if (name === NATIVE_WIN || name === null) {
+    // switching into the native context
+    this.curContext = null;
+    if (this.isRealDevice()) {
+      this.remote.disconnect();
+    }
+  } else {
+    // switching into a webview context
+    let idx = name.replace(WEBVIEW_BASE, '');
+    if (idx === WEBVIEW_WIN) {
+      // allow user to pass in "WEBVIEW" without an index
+      idx = '1';
+    }
+
+    // if contexts have not already been retrieved, get them
+    if (_.isUndefined(this.contexts)) {
+      await this.getContexts();
+    }
+
+    if (_.contains(this.contexts, idx)) {
+      let pageIdKey = parseInt(idx, 10);
+
+      if (!this.isRealDevice()) {
+        await this.remote.selectPage(pageIdKey, skipReadyCheck);
+        this.curContext = idx;
+      } else {
+        if (this.remote) {
+          await this.remote.disconnect();
+        }
+        this.curContext = idx;
+        await this.remote.connect(idx);
+      }
+    } else {
+      throw new errors.NoSuchContextError();
+    }
+  }
+
+  // attempt to start performance logging, if requested
+  if (this.perfLogEnabled && this.remote) {
+    logger.debug(`Starting performance log on '${this.curContext}'`);
+    this.logs.performance = new IOSPerformanceLog(this.remote);
+    this.logs.performance.startCapture();
+  }
+};
+
+commands.getWindowHandle = async function () {
+  if (!this.isWebContext()) {
+    throw new errors.NotImplementedError();
+  }
+  return this.curContext;
+};
+
+commands.getWindowHandles = async function () {
+  if (!this.isWebContext()) {
+    throw new errors.NotImplementedError();
+  }
+
+  let pageArray = await this.listWebFrames();
+  this.windowHandleCache = _.map(pageArray, this.massagePage);
+  let idArray = _.pluck(this.windowHandleCache, 'id');
+  // since we use this.contexts to manage selecting debugger pages, make
+  // sure it gets populated even if someone did not use the
+  // getContexts method
+  if (!this.contexts) {
+    this.contexts = idArray;
+  }
+  return idArray;
+};
+
+commands.setWindowHandle = async function (name, skipReadyCheck) {
+  if (!this.isWebContext()) {
+    throw new errors.NotImplementedError();
+  }
+
+  if (_.contains(_.pluck(this.windowHandleCache, 'id'), name)) {
+    let pageIdKey = parseInt(name, 10);
+    if (!this.isRealDevice()) {
+      await this.remote.selectPage(pageIdKey, skipReadyCheck);
+      this.curContext = this.curWindowHandle = name;
+    } else {
+      if (name === this.curWindowHandle) {
+        logger.debug(`Remote debugger is already connected to window '${name}'`);
+      } else if (_.contains(_.pluck(this.windowHandleCache, 'id'), name)) {
+        await this.remote.disconnect();
+        this.curContext = this.curWindowHandle = name;
+        await this.remote.connect(name);
+      } else {
+        throw new errors.NoSuchWindowError();
+      }
+    }
+  } else {
```

Comment:
hate to nitpick, but this is popping out at me as needing to go up top as well :-) just seems natural to have your chain of "if this isn't the case, then throw an error" up top, with the meat below